### ISSUE TITLE
Only use relevant values in calculations (e.g., like Collect)

### ISSUE THREAD

**Issue by yanokwa** [enketo/enketo-core#849]
In Collect, when a question is not relevant, its value is treated as blank for all expressions it is used in, including relevance. In Enketo, even irrelevant values are used in calculations. This difference between Collect and Enketo is unexpected and we think Enketo's behavior should change because it's the least widely deployed. 

This change will most certainly break forms in folks using Enketo outside of ODK core tools and so we do need to be cautious about how this is communicated and rolled out.

More discussion (including references to the vagueness of the XForms spec on this issue) are at https://forum.getodk.org/t/enketo-v3-uses-non-relevant-values-in-calculations-whereas-collect-does-not/35567/

**Comment by MartijnR** [comment]
> its value is treated as blank for all expressions it is used in

Does this mean that the value is set to empty in the model as soon as it becomes non-relevant?

Will try to find Enketo's reasoning fwiw. I believe it was done to not loose user-entered values if users accidentally make a group with entered data non-relevant (and because the XForms spec talks about "pruning" non-relevant values before submission somewhere if I remember correctly).

There is already a mechanism (search for `emptyNonRelevant` in [calculate.js](https://github.com/enketo/enketo-core/blob/master/src/js/calculate.js) to make this change. I think OpenClinica always clears non-user-entered values in questions with calculations when they become non-relevant, and it uses that `emptyNonRelevant` parameter to provide this customization.

related bug (in repeats): https://github.com/enketo/enketo-core/issues/830

**Comment by lognaturel** [comment]
> Does this mean that the value is set to empty in the model as soon as it becomes non-relevant?

No! It's part of the evaluation logic to check relevance and treat values that are non-relevant as blank. So if a value becomes relevant again, it does still have the user-input value. The pruning happens when a user leaves the current form editing session.

**Comment by MartijnR** [comment]
Ah, in that case ignore what I wrote above about the existing mechanism.

```xml
<data> 
   <a>2</a>
   <b>3</b>
</data>
```

So just to be sure, if a calculation `/data/a + /data/b` is sent to the XPath evaluator with the above model, and `/data/a` is non-relevant, the result would be `3`?

**Comment by lognaturel** [comment]
> concat(/data/a, data/b) would be "b"

I think you mean `3` and yes, that's right.

> /data/a + /data/b is sent to the XPath evaluator with the above model, and /data/a is non-relevant but /data/b is relevant, the result would be NaN

It's saved as blank/null but yes. We recommend the pattern `if(/data/a = '', 0, /data/a) + if(/data/b = '', 0, /data/b)` for that kind of thing.

[Here](https://docs.google.com/spreadsheets/d/1IBNqb8H6NZ_P_9KGyzS0z0kGgebTmMRjpz1bj-Qq-mk/edit?usp=sharing) is a form to play with.

**Comment by MartijnR** [comment]
Oops, yes I meant `"3"`. :face_with_spiral_eyes:

Since the XPath evaluator (Enketo's and the native evaluator in browsers) only really knows about the model and the expression (and context, resultType), this may be quite a challenge to implement in Enketo. Maybe you and @eyelidlessness have a cleverer idea, but I am thinking you'd have to send a pruned copy of the model to the evaluator. That would be a serious performance degradation.

In any case, it would be great if this change could be implemented with a configuration item. This will greatly help Enketo users that don't use ODK Collect (and would be more likely have issues with the change).

**Comment by eyelidlessness** [comment]
This seems like it may be another use case for [exposing the XPath evaluator's parsing functionality](https://github.com/enketo/enketo-core/issues/823#issuecomment-939153315), which enketo-core could in turn use to determine [ir]relevant sub-expressions in a calculation.

**Comment by pbowen-oc** [comment]
If you make this change, we would definitely ask that it be configurable. We don't hide or clear non-relevant values (unless they are calculated) so that the user doesn't mistakenly clear user-entered data due to relevance changes and have resulting changed values submit right away (with real-time item submissions). So, we need calculations to be maintained based on the current values of items.

**Comment by lognaturel** [comment]
Thanks for chiming in, @pbowen-oc. 

> We don't hide or clear non-relevant values

Can you say more about this? The relevance of a question is supposed to mean whether it appears on screen or not. Have you changed that? Let's say there are a number of questions only relevant to individuals over the age of 65, some of those questions were answered and then it was discovered the age was accidentally entered as 74 when it should have been 64. The age is corrected. What happens to the over 65 questions that are no longer relevant?

My hunch is that regardless of Collect and its behavior, Enketo's behavior in this area leads to subtle data errors or user experience issues. Some examples:
- A form first asks whether some conditions are relevant to a person and exposes questions for the conditions that are relevant. Then it computes some summary value based on answers in sections that may or may not have been relevant. If a value is entered and then the whole section is made non-relevant, with Enketo, that value is still included in the summary calculation. A concrete example: a household survey asks which monthly bills a family has (phone, internet, gas, electric, etc). For each bill, it asks a series of questions including how much is paid. Internet is selected and an estimated cost is entered. Then the respondent remembers they cancelled their internet subscription last month so that is unselected. The computed monthly bills sum would still include this value.
- A form describes a decision tree. Unless the value of each parent node in the tree is considered in relevance expressions, the wrong questions may be shown if answers get changed. A concrete example: a form represents a phone script. Did someone pick up y/n? Were you able to get the respondent on the line y/n? Does the respondent consent? y/n Given consent, the rest of the survey is shown. Let's say someone picks up. Then the caller thinks they have the right respondent. They ask for consent, the survey is shown, and they and start answering. Then they realize they have the wrong person. If the survey relevance expression is just `/data/consent = 'y'`, the survey will still show. This is pretty easy to catch and fix to `/data/respondent_available = 'y' and /data/consent = 'y'` but as the decision tree gets more complicated this is very difficult to get right. This is what the original forum post was about.

**Comment by pbowen-oc** [comment]
@lognaturel - For our use case with data being submitted in real time, we don't hide non-calculated values that become non-relevant. We have to show all data in its current state so that the user knows the data on the form matches what is saved in the database. We don't want the user to be unclear about whether the form represents the current data (which is, I think, similar to the angle you are coming at this from for your use case). 

We don't want the user to make a mistaken click that causes an item or group to become non-relevant and clear that data since it could lead to substantial data loss (since in our case, those null values would be submitted immediately). Our implementation is to maintain and display non-calculated values when they become non-relevant (along with an error message on those values) so that the user can make the appropriate update for their needs: either update the triggering data to make the items relevant again or clear the non-relevant values. Even outside of our autosave use case, since a click on one page of a form can make items on other pages non-relevant, it's unclear to me how users would even know that that they may have made a significant change to a form's data (i.e., clearing lots of it) with a mistaken click/tap if the result is to hide and effectively clear data with no warning.

For example, an item for "Were labs collected?" followed by many items to collect lab data as appropriate. After selecting that labs were entered and then entering specific lab data, if the user then indicates that labs weren't collected, we force them to update the data to a consistent state rather than assume that their final click indicating labs weren't entered was intentional and their prior data entry for lab values was erroneous. 

Also, with triggered calculations there seems to be additional danger that recalculating a value even when its source data has not changed could lead to unintended re-triggering of calculations. I'm not thinking of any specific case for this, rather just would could be problematic about it.

**Comment by jkuester** [comment]
I just wanted to add that this issue can cause very confusing behavior in forms with non-relevant fields that have default values.  For example:

type | name | label | calculation | default | relevant
-- | -- | -- | -- | -- | --
string | world | World |   | World | false()
calculate | world_calc | Hello Calc | string(${world}) |   |  
note | message | Hello: ${world_calc} |   |   |  

When running this form with Enketo, the user will see "Hello: World" as the `message`, but when the form is saved, the value for `world_calc` will be empty (since the non-relevant fields are [cleared upon validation](https://github.com/enketo/enketo-core/blob/b3b6fcdca7a7ce10eeed5d3a9ab8b9bb69e99145/src/js/form.js#L961)).

**Comment by lognaturel** [comment]
Indeed, @jkuester, that's very confusing. Do your users run into some of the chained relevance issues I mentioned above? 

This has become a priority for @eyelidlessness and me because users keep running into it in all kinds of ways. I assume that OpenClinica users either don't make forms with multiple levels of decisions or don't end up hearing about form design issues because form fillers kind of manage to get things in a consistent state in their form filling session. Or perhaps something about their forms means they don't change answers often within a form filling session.

Unfortunately, I still don't understand how OpenClinica would be negatively affected, @pbowen-oc. Changing the answer to e.g. "Were labs collected?" to "no" would indeed update calculations that rely on values made non-relevant. But if the answer were changed back to "yes", those calculations would be updated back to what they were previously. Seems like you'd still get the error message you show on newly non-relevant values.

As we start working on this we'll try to look for a reasonable way to make it configurable.

**Comment by jkuester** [comment]
@lognaturel I do not believe that config for cht-core can run into the chained relevance issues you mentioned above because we are still on `4.41.6` of enketo-core and run with `clearIrrelevantImmediately = true` which I think means that in _most_ cases the values are getting cleared anyway as soon as the question becomes irrelevant.

The reason this issue of non-relevant values in calculations has come up for us is that, while trying to get the CHT uplifted to a newer version of Enketo, I have found that some of our forms are broken because the behavior of non-relevant questions with `default` values changed.  The details of the situation can be read [here](https://github.com/medic/cht-core/issues/6345#issuecomment-1057466752), but the TLDR is that a user can fill out a form and save it successfully and the values that are actually stored for the response are totally different from what the user entered because everything was recalculated during the final validation without the non-relevant value.

At this point I am expecting that we will have to update the configuration for any forms that have this problem, but simply identifying forms with the problem has been exacerbated by this issue since you cannot see the problem with the logic until the form has been validated.  So, from that perspective, I think we would definitely prefer a solution to this issue where non-relevant values are treated as blank all the time (and not just when they are cleared out at the end).

**Comment by lognaturel** [comment]
> the behavior of non-relevant questions with default values changed.

You mean that because you no longer have the option of using `clearIrrelevantImmediately = true`, you're running into the problem discussed in this issue, right?

> I am expecting that we will have to update the configuration for any forms that have this problem

What do you mean by configuration in this context? You mean change the form design?

We're motivated to get this addressed soon. Perhaps you could help by adding some tests, @jkuester (probably once there's an initial implementation)?

**Comment by jkuester** [comment]
> You mean that because you no longer have the option of using `clearIrrelevantImmediately = true`, you're running into the problem discussed in this issue, right?

@lognaturel Yes, although technically it was the `clearIrrelevantImmediately = true` plus some logic in the `relevants.disable` function to keep the default values [that has also been removed](https://github.com/enketo/enketo-core/commit/1a32cd52ef098c86d906876dd5d538f7979067d1).

> What do you mean by configuration in this context? You mean change the form design?

Yes. Frankly the form having logic based on the default value of a non-relevant field seems pretty sketchy to begin with, so refactoring the form design feels like the way forward. At this point I have not been able to think of any viable alternatives that would allow for maintaining this functionality (even if we wanted to).

I am definitely happy to do what I can to help move forwards on a solution here (especially now that I am well acquainted with some of the behaviors at play...)!

**Comment by jkuester** [comment]
Minor, but after thinking about this some more, I am pretty sure that the form refactoring that I described above can be very minor.  Basically I just need to replace references to `non_relevant_question` with references to a `calculate` that is always relevant with the calculation: `coalesce(${non_relevant_question}, *original default for non_relevant_question*)`.  This way, in the cases where the question is relevant, its value will take precedence, but otherwise we still get the desired default value.  So, the fix seems simple, but actually knowing the form is broken is still the harder part....

**Comment by lognaturel** [comment]
Agreed the form design change you described would get to the desired behavior as I understand it. However, I think it's very likely that you have forms with relevance issues. These tend to manifest when changing responses so they can also be hard to track down. Even knowing the way Enketo works with chained relevance I find it very difficult to write correct relevance conditions. I now typically use the XLSForm `trigger` column to clear a value when the question its relevance is based on changes but that doesn't work with relevance expressions that reference multiple questions.

**Comment by lognaturel** [comment]
Context from @MartijnR about why `clearIrrelevantImmediately` was removed: https://github.com/enketo/enketo-core/commit/26fb7d47921a71a1af75e95e4e75f1556a6d7c6d#commitcomment-68248306 @jkuester do you know why y'all were using it? To be able to author decision tree types of forms? To more closely match Collect's behavior? Something else?

**Comment by jkuester** [comment]
@lognaturel, `clearIrrelevantImmediately = true` is the default config in `4.41.6`, and we were just inheriting that default value in the CHT config. It may not have ever been a conscious decision.  That being said, our deployments have a large number of "decision tree"-style forms with very complex (at least to me :smile: ) nested logic. So, I expect that we would have chosen to use `clearIrrelevantImmediately = true` if its default value had been false....

**Comment by pbowen-oc** [comment]
@lognaturel - Sorry for the delay in responding to this thread. Regarding the way our data entry users use forms, I think it is probably true that they don't typically make many data changes within any given form editing session. Since they are capturing clinical data, they shouldn't be changing an item to a different value within a form session unless there they make a mistake. Of course, it still does happen many times a day given the volume of studies run on OpenClinica. Some forms are quite simple, but it is common for forms to have many logical pathways and our customers continually push the boundaries of what size and complexity of a form is useable (which is why we needed the new evaluator). If A is relevant based on B, and B is itself relevant based on C, in most cases our users would simply write relevant logic for A based on B and not factor C into it. I have seen a small set of cases where the logic for A specifically includes B and C, but that is not the standard way it is handled. Our help desk would know if there were data issues since those would always be escalated to them if they occurred since data issues are the most important thing customer data managers are lookin for.

At a high level, the behavior discussed would not work for us since it would lead to inconsistencies in form data. A calculation that is a function of A and B would not be guaranteed to have a value that is derived from the current values of A and B (since one or both of them could be non-relevant). This could lead to confusion within the form session for a user that sees the inconsistent values A, B, and the calculation. As you noted, if the user updates the source data so that A and B become relevant again, the calculation should return to the previous state. In the meantime, however, the calculation was submitted to the database with a different value and triggers an unexpected change in the audit log. If the user closes the form without resolving the relevance error, the calculated value will remain incorrect in the database until the user returns to fix it or a data manager catches the issue during data cleaning. The incorrect data will impact interim data reports and other form or system logic that utilize this data. 

Suppose there is a question on a visit form for "Was blood pressure measured?". If it is Yes, then the user can enter Systolic and Diastolic pressures. The form could calculate mean arterial pressure and a qualitative rating of whether the pressure is high, normal, or low. Those calculations might not be displayed on the form at all, but just saved in the background. Suppose the data entry user selects Yes, enters 180 over 120 for pressure, and at some point mistakenly unselects Yes. If this were to save the calculated mean pressure and rating to the database based on Systolic and Diastolic being null, then reports for the clinician and study participant would not be able to accurately show that this participant had a high blood pressure reading at this visit. System logic that automatically schedules a blood pressure management form to ensure that this high measurement is addressed wouldn't fire. Similarly, if the next visit form pulls this data in and it should cause blood pressure management questions to be relevant due to the previous high rating, that wouldn't happen as expected. If the data entry user addresses the relevance error on the form quickly, it will be less likely to have more significant impacts elsewhere, but if they leave the form without addressing the error, the erroneous calculations can have broad effects.

Staying within the form, if the calculated item is a trigger for another item's triggered calculation, then the change in value to the calculation will have unintended effects on that other item even if the user immediately corrects their mistake and restores the items to relevant. 

Suppose, the blood pressure rating calculation is a trigger for another item and is used to calculate a sort of default treatment plan that the data entry user can modify if needed. Suppose the user enters a normal blood pressure and this triggers an item later in the form to default to no blood pressure treatment needed, but the data entry user manually changes this to treatment needed due to other factors too complex to include in that triggered calculation. If the user momentarily unselects Yes and selects it again right away, the behavior discussed would cause the treatment item later in the form to return to the no treatment needed state. The data entry user could be unlikely to think of checking this since they hadn't changed the blood pressure values and didn't see a change in the blood pressure values.

I think the difference in the behavior you described wanting and the behavior I am describing comes down to the same general reasoning. In your case, the non-relevant item is effectively null (it can be restored with its prior value, but it will be saved as null if the form is submitted at that time). In our case the non-relevant item still has its value unless the user specifically clears it.

I am not sure where you are at on this based on the most recent comments here.

**Comment by MartijnR** [comment]
> I am curious to understand how you are currently achieving your desired behavoir.

Their fork has a bunch of customizations including [this](https://github.com/OpenClinica/enketo-express-oc/blob/master/public/js/src/module/form.js#L159). They actually clear non-user-entered non-relevant (calculated) values immediately, and never clear user-entered values.

**Comment by lognaturel** [comment]
> clearIrrelevantImmediately = true is the default config in 4.41.6

That's very useful context, thanks. That may explain why some users could have sworn their forms used to work correctly (though it would have been a long time ago).

> our deployments have a large number of "decision tree"-style forms

@jkuester it would help if you could identify 2-3 forms from your corpus that behave badly with the current `enketo-core` version. I think you'll find that chained relevance in decision trees cause problems when a value is changed and it would be good to confirm.

> This could lead to confusion within the form session for a user that sees the inconsistent values A, B, and the calculation.

That makes sense, and it's because you continue to show non-relevant values. For stock Enketo, the current behavior is confusing for a similar reason -- values that are invisible and eventually will be cleared are used in expressions. You show and save non-relevant values in the initial form filling session, right, not just in an editing session?

> I am not sure where you are at on this

We are planning to make this change and trying to give you a config flag. Our first spike did not address all of our test cases so we're trying a different approach. We'll spend some time on QA before releasing.

**Comment by MartijnR** [comment]
> Can you please share an abridged example? I can't really tell whether the undesirable behavior is an interaction with your customizations or not.

Sorry, I was going to provide a summary but forgot. @pbowen-oc, please correct if necessary. It's not related to an OpenClinica customization.

The case includes a setvalue/xforms-value-changed (dynamic default) directive that provides a default treatment value to a 'treatment' question. This value can be overridden to a custom treatment answer by the user/doctor. If the user accidentally makes this question non-relevant, that overridden value would be cleared forever. If the user corrects the relevancy, the setvalue/xforms-value-changed default will come back.

So e.g. the 'treatment' dynamic default value is 'no treatment needed', but the doctor manually changes this value to 'followup and additional lab testing required'. After the non-relevancy/relevancy switch, the value will be 'no treatment needed' again.

> We are planning to make this change and trying to give you a config flag.

Thank you!

**Comment by jkuester** [comment]
> @jkuester it would help if you could identify 2-3 forms from your corpus that behave badly with the current enketo-core version. I think you'll find that chained relevance in decision trees cause problems when a value is changed and it would be good to confirm.

@lognaturel, a good example of where we see issues is the `immunization_visit` form in [this directory](https://github.com/medic/cht-core/tree/master/config/standard/forms/app).  The value for the `source` field can get set to "task" via the input data when the form is initially opened, but otherwise defaults to "user".  The group that contains the `attended_imm` field is only relevant if `${source} = 'task'`. (And the default value of `attended_imm` is 'yes'.)  All the rest of the form questions (contained in the `group_select_vaccines` group) are only relevant when `${attended_imm} = 'yes'`.

So, if an `immunization_visit` form is manually created by the user (and not opened from a task), the `attended_imm` question is never relevant (since `source` is not "task").  Due to this issue that we are currently discussing, the user still sees the rest of the questions (in the `group_select_vaccines` group) and can continue to fill out the form, recording the vaccines that were received.  However, when the user saves the form and the CHT triggers `Form.validateAll` before recording the data, the default "yes" value for `attended_imm` is cleared which causes everything in `group_select_vaccines` to become irrelevant and be cleared. So the resulting data that is stored for the form indicates that no vaccines were received.

Very similar patterns are also used in the `delivery`, `nutrition_followup`, `postnatal_visit`, and `pregnancy_visit` forms in that same directory that I have linked above.

**Comment by lognaturel** [comment]
Thanks, @jkuester, that's helpful.

@MartijnR, @pbowen-oc I've tried writing an example form [here](https://docs.google.com/spreadsheets/d/1dLaf3dQmMo3yk0iIaouUTl7rPAEcmHY2bYBWjzTVi0Q/edit#gid=0) but I don't think I'm quite getting it. It's hard for me to think of what a good trigger would be in a case like this where multiple values are involved. I agree that it's hard to reason about triggers and relevance together and it would help to have some concrete examples to write tests against.

A long time ago we'd talked about OpenClinica adding tests on forms that are important to them. That would certainly help.

**Comment by MartijnR** [comment]
Thanks for doing that @lognaturel! And sorry, I should have done that myself. I think this form represents the case correctly. I just changed the question type of the last question to text.
 
So after responding "yes", "130" and "90", the dynamic default for treatment would be "default". If the user changes that default that with "immediately administer some drugs to keep this patient alive", that custom text would disappear forever if the diastolic value is automatically cleared (after pb_taken was accidentally changed to 'no').

I agree we definitely should write a test for this. I can do this in a PR after @pbowen-oc confirms this form accurately represents an issue (if we didn't have a configuration item).

**Comment by lognaturel** [comment]
Would be great to get the real shape of the form. In what I did, update is only triggered by `diastolic`. So if someone enters 109 and 90 but then realizes they made a typo and it should be 190, the recommended treatment is still none.

**Comment by eyelidlessness** [comment]
This is implemented in #870 and ready to be tested! The default behavior remains the same so it won't be a breaking change, but the new behavior can be enabled in `config.js` as described in the PR notes.

**Comment by lognaturel** [comment]
@jkuester have you had a chance to take a look at the behavior in #870? Any tests you want to make sure are added?

@joshuaberetta @jnm @ukanga I want to make sure that you have seen that we are working on aligning how relevance is treated with Collect. Perhaps you have test forms you'd like to make sure are considered? Forms with cascading relevance (e.g. decision trees) are the most likely to be affected. Once this gets merged it will require a config flag in Express.

**Comment by jkuester** [comment]
@lognaturel sorry for the delayed response here!  It took me some time, but I have gone through #870 and even tried spinning it up in cht-core with `excludeNonRelevant: true`.  Everything looks good!  The behavior is a big improvement for us since we can now properly see, while filling out forms, the same calculated data that will ultimately be saved (and not have data "magically" change on us when saving because non-relevant fields with default values get cleared).

(Of course, I also noticed that is will break some of our current "hacks" that we to keep a top-level `inputs` group when saving (regardless of its relevancy), but stuff like that is expected and we really need to come up with a better way of handling that in the CHT anyway...)

I have not really been able to think of too many test cases to add beyond all the ones that @eyelidlessness already has! (Super great to see all of those!)  The only thing was that I did not see any existing cases that involved a field with a default value. So, I created this PR that updates a few tests to cover that behavior: https://github.com/eyelidlessness/enketo-core/pull/1

**Comment by lognaturel** [comment]
This is wonderful, @jkuester!! Thanks a bunch for the added test cases.

> keep a top-level inputs group when saving (regardless of its relevancy)

Can you describe the semantics of that group? e.g. does it represent device state? Actions a user took before reaching a form? ...?

**Comment by jkuester** [comment]
> Can you describe the semantics of that group and how it's populated? e.g. does it represent device state? Actions a user took before reaching a form? ...?

Basically, 'yes' to both. In [our documentation](https://docs.communityhealthtoolkit.org/apps/reference/forms/app/#xform) we describe the `inputs` group as:

> Data needed during the completion of the form (eg patient’s name, prior information) is passed into the inputs group.

So, for all forms, the `inputs` data contains information about the patient associated with the current form and the user submitting the form. But in some situations the `inputs` data can be a lot more in-depth and include things like historical data from previous forms, etc. Generally speaking, we want to include the `inputs` data when saving the form. However, this is where things get nasty. Somewhere along the way a patter was introduced which allowed for including actual questions in the `inputs` group. These questions might in some cases have answers supplied already by the system (in which case the `inputs` group is set to non-relevant so the user cannot see the questions). But, if the user needs to answer the questions in the `inputs` group, the group is relevant and visible to the user.  In either case, though, when saving the form we want to include the `inputs` data (whether it is relevant or not). (In a way, I think this behavior is similar to the challenges mentioned with the OpenClinica workflows above where relevance is being used to toggle visibility of questions when we actually still want to record the data from them...) Practically speaking, we populate the `inputs` data by injecting it directly into the `instanceStr` xml that we pass in when constructing the Enketo `Form` object.

Honestly, the more I think about this, the more I am worried that we have backed ourselves into a corner here.  :worried:  Previously, we were able to tweak things into working since we could reference the `inputs` data while the form was being filled out regardless of the relevance. Then, to keep the data at the end, when saving, we just had to patch a bit of code to make sure the `inputs` was not cleared and was included in the call to `getDataStr`.  With these new changes, though, we can't just force `inputs` to be relevant up-front since that might show questions to the user that they are not supposed to see. But, if `inputs` is not relevant, then we will not be able to reference the data while filling out the form.

I am going to have to do some more thinking here.  It seems like I need to find a way to control question visibility based on the result of xPath logic but without affecting the question's relevance....

> Do you have any particularly complex forms you use to evaluate performance?

I don't have any off the top of my head, but I will ask around and see if I can come up with any that I can share!

**Comment by jkuester** [comment]
@lognaturel for the record, [this form](https://github.com/medic/cht-conf-test-harness/blob/master/test/collateral/forms/app/dtree-infant-child.xml) is the biggest (in both size and complexity) CHT form I have found that is publicly accessible.  When I get some time I can try and wire it up to a new test case so maybe we can evaluate the performance, but it will be a bit before I can get that done.

**Comment by eyelidlessness** [comment]
> With these new changes, though, we can't just force inputs to be relevant up-front since that might show questions to the user that they are not supposed to see. But, if inputs is not relevant, then we will not be able to reference the data while filling out the form.

I think this can still work with `excludeNonRelevant`, with a bit of indirection. You could reference the data in model nodes with no associated form controls, which can remain relevant and would be retained in submissions, and add model nodes to calculate/reference the same data for question inputs with their own relevance expressions. This would separate the presence of data in the model from whether it's presented to users in the form UI. Do you think that's an approach that would work for your use case?

**Comment by jkuester** [comment]
@eyelidlessness, yes, after finally getting some time to think about this more I believe what you are proposing should work in principal (and as you say, would also solve our existing challenges around saving `inputs` data too).  

How to actually do that in practice might be more complex :smile: since ideally we would want to make it passive (somehow) to existing xml form configs. But, maybe a bit of clever xslt could do the trick...
